### PR TITLE
For exact checks use `is` / `is not`

### PR DESCRIPTION
**Pull request by tkoyama010**
### Overview

Fix error #4724

### Details

- None

### CODE REVIEW

File: pyvista/core/filters/data_set.py [hunk 1]
```diff
@@ -4906,7 +4906,7 @@ def merge(
         _update_alg(append_filter, progress_bar, 'Merging')
         merged = _get_output(append_filter)
         if inplace:
-            if type(self) == type(merged):
+            if isinstance(self, type(merged)):
                 self.deep_copy(merged)
```
Comment (MatthewFlamm) [inline pyvista/core/filters/data_set.py:4910]: The original logic seems to be that the types have to be the same.  The error message in this case says to use `is` comparison in this case.  I haven't checked whether this change makes more sense though.

Comment (tkoyama010) [reply pyvista/core/filters/data_set.py:4910]: Thanks for your comment, I am looking into it as I don't understand what the difference is between the two methods.

Comment (tkoyama010) [reply pyvista/core/filters/data_set.py:4910]: Ah, I see. [isinstance](https://docs.python.org/3/library/functions.html#isinstance) includes the inherited class as well, so a comparison by is is appropriate, as you say.

File: pyvista/core/filters/data_set.py [hunk 2]
```diff
@@ -4906,7 +4906,7 @@ def merge(
         _update_alg(append_filter, progress_bar, 'Merging')
         merged = _get_output(append_filter)
         if inplace:
-            if type(self) == type(merged):
+            if isinstance(self, type(merged)):
```
Comment (tkoyama010) [inline pyvista/core/filters/data_set.py:4909]:
```suggestion
            if type(self) is type(merged):
```

File: tests/core/test_datasetattributes.py [hunk 3]
```diff
@@ -436,7 +436,7 @@ def test_length_should_be_0_on_clear(insert_arange_narray):
 def test_keys_should_be_strings(insert_arange_narray):
     dsa, sample_array = insert_arange_narray
     for name in dsa.keys():
-        assert type(name) == str
+        assert isinstance(name, str)
```
Comment (tkoyama010) [inline tests/core/test_datasetattributes.py:439]:
```suggestion
        assert type(name) is str
```

File: tests/core/test_datasetattributes.py [hunk 4]
```diff
@@ -447,7 +447,7 @@ def test_key_should_exist(insert_arange_narray):
 def test_values_should_be_pyvista_ndarrays(insert_arange_narray):
     dsa, sample_array = insert_arange_narray
     for arr in dsa.values():
-        assert type(arr) == pyvista.pyvista_ndarray
+        assert isinstance(arr, pyvista.pyvista_ndarray)
```
Comment (tkoyama010) [inline tests/core/test_datasetattributes.py:450]:
```suggestion
        assert type(arr) is pyvista.pyvista_ndarray
```

**Review by MatthewFlamm — APPROVED** [review]
LGTM

**Comment by bjlittle** [comment]
Wouldn't pyupgrade catch these types of changes in syntax usage?

**Comment by tkoyama010** [comment]
@bjlittle 
None of the [implemented features](https://github.com/asottile/pyupgrade#implemented-features) of pyupgrade seem to apply :)

**Comment by bjlittle** [comment]
> @bjlittle 
> None of the [implemented features](https://github.com/asottile/pyupgrade#implemented-features) of pyupgrade seem to apply :)

And ruff?

**Comment by MatthewFlamm** [comment]
It was _caught_ by `pyupgrade`, but I'm not sure if it would autofix given the likelihood that the code usage wasn't precise.

**Review by adeak — APPROVED** [review]
LGTM too, thanks @tkoyama010.

For anyone wondering, it took me a while to understand why this even matters. In most cases there's no difference, because if you do `class Foo: pass` then `Foo.__eq__ is object.__eq__`, in other words equality comparison falls back to object identity.

So you can only make `type(foo) == type(bar)` different from `type(foo) is type(bar)` if you use a metaclass. Here's a silly example in case anyone's interested:
```py
class SillyMeta(type):
    def __eq__(cls, other):
        return True

class Foo:
    pass

class Bar(metaclass=SillyMeta):
    pass

print(Bar == Foo)  # True
```

**Comment by tkoyama010** [comment]
@adeak Thanks. It's an art of bug.

**Pull request merged**